### PR TITLE
External transition type level validation

### PR DESCRIPTION
**Pull request by nholland94**
There are some long term upgrades I want to make to this mechanism, however, in the interest of time, I am going to call this good enough to merge into master. I plan to improve the interface with some more type aliases (and hopefully better names), and have a tighter interface that exposes less information. I made a couple of issues to track the remaining work.

https://github.com/CodaProtocol/coda/issues/2479
https://github.com/CodaProtocol/coda/issues/2480

### CODE REVIEW

File: src/lib/verifier/prod.ml [hunk 1]
```diff
@@ -127,7 +115,8 @@ end
 
 type t = Worker.Connection.t
 
-let create ~conf_dir =
+(* TODO: investigate why conf_dir wasn't being used *)
```
Comment (cmr) [inline src/lib/verifier/prod.ml]: The verifier ought to be stateless, not sure why it would need the conf_dir.

Comment (nholland94) [reply src/lib/verifier/prod.ml]: Yeah I just left it when I went through but it's unused. I'll just remove it.

File: src/lib/transition_frontier_persistence/transition_frontier_persistence.ml [hunk 2]
```diff
@@ -144,22 +138,30 @@ module Make (Inputs : Intf.Main_inputs) = struct
...
+  let directly_add_breadcrumb ~logger ~verifier ~trust_system
+      transition_frontier transition parent =
     let log_error () =
       Logger.fatal logger ~module_:__MODULE__ ~location:__LOC__
-        ~metadata:
-          [("hash", State_hash.to_yojson (With_hash.hash transition_with_hash))]
+        ~metadata:[("hash", State_hash.to_yojson (With_hash.hash transition))]
         "Failed to add breadcrumb into $hash"
     in
+    (* TODO: TMP HACK: our transition is already validated, so we "downgrade" it's validation *)
```
Comment (cmr) [inline src/lib/transition_frontier_persistence/transition_frontier_persistence.ml]: reminder that this is here. what's TODO here? a variant of `Breadcrumb.build` that takes a fully validated transition?

File: src/lib/transition_handler/processor.ml [hunk 3]
```diff
@@ -11,43 +11,152 @@ open Async_kernel
...
+            ~frontier initially_validated_transition
+        with
+        | Ok t ->
+            return (Ok t)
+        | Error `Not_selected_over_frontier_root ->
+            let%map () =
+              Trust_system.record_envelope_sender trust_system logger sender
+                ( Trust_system.Actions.Gossiped_invalid_transition
+                , Some
+                    ( "$state_hash was not selected of transition frontier root"
```
Comment (cmr) [inline src/lib/transition_handler/processor.ml]: `s/of/over`

File: src/lib/transition_handler/processor.ml [hunk 4]
```diff
@@ -11,43 +11,152 @@ open Async_kernel
...
+              "refusing to process $state_hash because is is already in the \
+               transition frontier" ;
+            return (Error ())
+        | Error `Parent_missing_from_frontier ->
+            Catchup_scheduler.watch catchup_scheduler
+              ~timeout_duration:catchup_timeout_duration
+              ~cached_transition:cached_initially_validated_transition ;
+            return (Error ())
+      in
+      (* TODO: only access parent in transition frontier once (already done in call to validate dependencies) *)
```
Comment (cmr) [inline src/lib/transition_handler/processor.ml]: can you open an issue for this?

File: src/lib/transition_handler/processor.ml [hunk 5]
```diff
@@ -11,43 +11,152 @@ open Async_kernel
...
+        cached_transform_deferred_result cached_initially_validated_transition
+          ~transform_cached:(fun _ ->
+            Transition_frontier.Breadcrumb.build ~logger ~verifier
+              ~trust_system ~sender:(Some sender) ~parent:parent_breadcrumb
+              ~transition:mostly_validated_transition )
+          ~transform_result:(function
+            | Error (`Invalid_staged_ledger_hash error)
+            | Error (`Invalid_staged_ledger_diff error) ->
+                Logger.error logger ~module_:__MODULE__ ~location:__LOC__
+                  ~metadata "error while building breadcrumb in processor: %s"
```
Comment (cmr) [inline src/lib/transition_handler/processor.ml]: Error in metadata please

File: src/lib/transition_handler/processor.ml [hunk 6]
```diff
@@ -132,84 +223,18 @@ module Make (Inputs : Inputs.With_unprocessed_transition_cache.S) :
...
-                           Logger.error logger ~module_:__MODULE__
-                             ~location:__LOC__
-                             "error while adding transition: %s"
-                             (Error.to_string_hum err) ) ) ) ))
+                   | Ok () ->
+                       ()
+                   | Error err ->
+                       Logger.error logger ~module_:__MODULE__
+                         ~location:__LOC__
+                         "failed to attach breadcrumb proposed internally to \
```
Comment (cmr) [inline src/lib/transition_handler/processor.ml]: error in metadata

File: src/lib/transition_router/initial_validator.ml [hunk 7]
```diff
@@ -1,80 +1,93 @@
...
-   and type external_transition_verified :=
-              Inputs.External_transition.Verified.t = struct
-  open Inputs
+  let handle_validation_error ~logger ~trust_system ~sender ~state_hash
+      (error : validation_error) =
+    match error with
+    | `Verifier_error err ->
+        Logger.error logger ~module_:__MODULE__ ~location:__LOC__
+          ~metadata:[("state_hash", State_hash.to_yojson state_hash)]
+          "Error while verifying blockchain proof for $state_hash: %s"
```
Comment (cmr) [inline src/lib/transition_router/initial_validator.ml]: error in metadata

File: src/lib/transition_router/initial_validator.ml [hunk 8]
```diff
@@ -1,80 +1,93 @@
...
+          (Error.to_string_hum err) ;
+        return ()
```
Comment (cmr) [inline src/lib/transition_router/initial_validator.ml]: This should proooobably be a `Trust_system.Actions.Sent_invalid_proof`? Crashing the verifier isn't that different than an actual invalid proof from this node's perspective...

File: src/lib/transition_router/initial_validator.ml [hunk 9]
```diff
@@ -1,80 +1,93 @@
...
+          (Trust_system.Actions.Gossiped_future_transition, None)
+    | `Invalid_time_received (`Too_late slot_diff) ->
+        Trust_system.record_envelope_sender trust_system logger sender
+          ( Trust_system.Actions.Gossiped_old_transition slot_diff
+          , Some
+              ( "off by $slot_diff slots"
+              , [("slot_diff", `String (Int64.to_string slot_diff))] ) )
+    | `Invalid_proof ->
+        Trust_system.record_envelope_sender trust_system logger sender
+          (Trust_system.Actions.Gossiped_invalid_transition, None)
```
Comment (cmr) [inline src/lib/transition_router/initial_validator.ml]: Likewise, `Sent_invalid_proof` is more specific here.

**Comment by cmr** [comment]
Looks good minus nits. I really like the way the validation code looks with this change.

File: src/lib/transition_router/initial_validator.ml (hunk 8 above)
Comment (nholland94) [reply src/lib/transition_router/initial_validator.ml]: fair enough

File: src/lib/coda_transition/external_transition.ml [hunk 10]
```diff
@@ -166,14 +120,366 @@ module Make (Staged_ledger_diff : Staged_ledger_diff_intf) :
...
+    let validate_frontier_dependencies (t, validation) ~logger ~frontier =
+      let open Result.Let_syntax in
+      let hash = With_hash.hash t in
+      let protocol_state = protocol_state (With_hash.data t) in
+      let parent_hash = Protocol_state.previous_state_hash protocol_state in
+      let root_protocol_state =
+        Transition_frontier.root frontier
+        |> Transition_frontier.Breadcrumb.transition_with_hash
+        |> With_hash.data
+        (* TODO: remove verified when this is plugged in *)
```
[the remark this answers is not in the document]
Comment (nholland94) [reply src/lib/coda_transition/external_transition.ml]: nope

File: src/lib/transition_handler/processor.ml (hunk 4 above)
Comment (nholland94) [reply src/lib/transition_handler/processor.ml]: https://github.com/CodaProtocol/coda/issues/2485

add issue # to comment as well

File: src/lib/transition_frontier_persistence/transition_frontier_persistence.ml (hunk 2 above)
Comment (nholland94) [reply src/lib/transition_frontier_persistence/transition_frontier_persistence.ml]: Removed the TODO and left the hack, but made an issue and comented with the number.
https://github.com/CodaProtocol/coda/issues/2486

**Pull request merged**